clients/src/main/java/org/apache/kafka/common/protocol/types/Schema.java

```diff
@@ -64,14 +64,23 @@ public void write(ByteBuffer buffer, Object o) {
     }
 
     /**
-     * Read a struct from the buffer
+     * Read a struct from the buffer. Missing fields at the end of the buffer are replaced with
+     * their default values if such fields are optional; otherwise a {@code SchemaException} is
+     * thrown to signify that mandatory fields are missing.
      */
     @Override
     public Struct read(ByteBuffer buffer) {
         Object[] objects = new Object[fields.length];
         for (int i = 0; i < fields.length; i++) {
             try {
-                objects[i] = fields[i].def.type.read(buffer);
```

Comment:
Are there other usages of `Type#read` which need to guard against underflow?

Reply:
Good question @mumrah ! I think I had checked when I added the fix, but I checked again. 

An similar check for underflow does not apply to the other implementations of `Type#read`. Specifically: 

* `ArrayOf` is accompanied by the size of the array of object that will be read, and having an array with some objects missing should not be supported. 
* All the primitive types have fixed size, so the type is either there as a whole or underflow is a valid exception. 
* Then, all the variable length types such as STRING, NULLABLE_STRING, BYTES, NULLABLE_BYTES, , VARINT, VARLONG, 
* Finally, RECORDS is just a wrapper to the byte buffer, so read just wraps the buffer and actual reading is left to the classes that use this type.